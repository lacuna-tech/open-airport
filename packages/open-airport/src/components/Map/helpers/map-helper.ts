import calculateDistance from '@turf/distance'
import { WebMercatorViewport } from 'react-map-gl'
import { MapViewportState } from 'components/Map/types'
import { Dimensions } from 'components'
import { LngLatBounds, LngLatBoundsLike } from 'mapbox-gl'

export const fitBounds: (options: {
  dimensions: Dimensions
  bounds: LngLatBoundsLike
  padding: number | number[]
  maxZoom: number
}) => Pick<MapViewportState, 'latitude' | 'longitude' | 'zoom'> = ({ dimensions, bounds, padding, maxZoom }) => {
  const { width } = dimensions
  // WebMercatorViewport's Bounds does not play well with Mapbox's Bounds, so a crude
  // array composition is needed.
  const coordinates = LngLatBounds.convert(bounds).toArray()
  const webMercatorBounds: [[number, number], [number, number]] = [
    [coordinates[0][0], coordinates[0][1]],
    [coordinates[1][0], coordinates[1][1]]
  ]
  const { latitude, longitude, zoom } = new WebMercatorViewport(dimensions).fitBounds(webMercatorBounds, {
    // Note: This can crash when padding is too large. E.g., left+right >= width, or top+bottom >= height.
    // You can more easily encounter crashes like this if chrome dev-tools is open pretty wide. The solve for
    // this is to use zoom scaled padding.
    padding: Array.isArray(padding)
      ? {
          top: padding[0],
          right: Math.min(padding[1], width - padding[3] - 1),
          bottom: padding[2],
          left: padding[3]
        }
      : padding
  })

  return { latitude, longitude, zoom: Math.min(zoom, maxZoom) }
}

export const getDistance = (
  start: { latitude: number; longitude: number },
  end: { latitude: number; longitude: number }
) =>
  calculateDistance(
    {
      type: 'Feature',
      properties: {},
      geometry: {
        type: 'Point',
        coordinates: [start.longitude, start.latitude]
      }
    },
    {
      type: 'Feature',
      properties: {},
      geometry: {
        type: 'Point',
        coordinates: [end.longitude, end.latitude]
      }
    }
  )
