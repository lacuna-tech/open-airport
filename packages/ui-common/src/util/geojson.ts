import { BBox, Feature, FeatureCollection, GeoJsonProperties, Geometry } from 'geojson'
import type { LngLatBoundsLike } from 'mapbox-gl'

export const isFeatureCollection = <G extends Geometry, P = GeoJsonProperties>(
  obj: Feature<G, P> | FeatureCollection<G, P>
): obj is FeatureCollection<G, P> => {
  return 'features' in obj
}

export const firstFeature: <G extends Geometry, P = GeoJsonProperties>(
  geography_json: Feature<G, P> | FeatureCollection<G, P>
) => Feature = geography_json => {
  return isFeatureCollection(geography_json) ? geography_json.features[0] : geography_json
}

export const bboxToLngLatBounds = (bounds: BBox): LngLatBoundsLike =>
  bounds.slice(0, 4) as [number, number, number, number]
