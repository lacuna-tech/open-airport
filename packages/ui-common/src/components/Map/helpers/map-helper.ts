import { WebMercatorViewport } from 'react-map-gl'
import { LngLatBounds, LngLatBoundsLike } from 'mapbox-gl'
import bbox from '@turf/bbox'
import circle from '@turf/circle'
import { Geography } from '@mds-core/mds-types'
import { Feature, FeatureCollection, GeoJsonProperties, Point, Polygon, MultiPolygon } from 'geojson'
import { interpolateRgb } from 'd3-interpolate'
import collect from '@turf/collect'
import * as simpleStats from 'simple-statistics'
import * as d3Array from 'd3-array'
import * as meta from '@turf/meta'
import { CommonConfig } from '@lacuna/agency-config'
import { Dimensions, MapViewportState } from '../types'

const {
  mapBox: {
    options: { fitBoundsMaxZoom }
  }
} = CommonConfig

export const fitBounds: (options: {
  dimensions: Dimensions
  bounds: LngLatBoundsLike
  padding: number | number[]
  maxZoom: number
}) => {
  latitude: number
  longitude: number
  zoom: number
} = ({ dimensions, bounds, padding, maxZoom }) => {
  const { width } = dimensions
  // WebMercatorViewport's Bounds does not play well with Mapbox's Bounds, so a crude
  // array composition is needed.
  const coordinates = LngLatBounds.convert(bounds).toArray()
  const webMercatorBounds: [[number, number], [number, number]] = [
    [coordinates[0][0], coordinates[0][1]],
    [coordinates[1][0], coordinates[1][1]]
  ]
  const { latitude, longitude, zoom } = new WebMercatorViewport(dimensions).fitBounds(webMercatorBounds, {
    // Note: This can crash when padding is too large. E.g., left+right >= width, or top+bottom >= height.
    // You can more easily encounter crashes like this if chrome dev-tools is open pretty wide. The solve for
    // this is to use zoom scaled padding.
    padding: Array.isArray(padding)
      ? {
          top: padding[0],
          right: Math.min(padding[1], width - padding[3] - 1),
          bottom: padding[2],
          left: padding[3]
        }
      : padding
  })

  return { latitude, longitude, zoom: Math.min(zoom, maxZoom) }
}
interface GeographyFeatureFromGeojsonOptions {
  /**
   * Number of verticies to give circular polygons generated from point geographies. Defaults to 64.
   */
  pointGeometryCircleResolution?: number
}

export const geographyFeatureFromGeojson = (
  geographyJson: FeatureCollection,
  options?: GeographyFeatureFromGeojsonOptions
): FeatureCollection<Polygon | MultiPolygon> => ({
  type: 'FeatureCollection',
  features: geographyJson.features.map(feat => {
    if (feat.geometry.type !== 'Point') {
      return feat as Feature<Polygon | MultiPolygon>
    }
    return circle(feat.geometry, 100, { units: 'feet', steps: options?.pointGeometryCircleResolution ?? 64 })
  })
})

export const getGeographyBounds: (options: { geography: Pick<Geography, 'geography_json'> }) => LngLatBoundsLike = ({
  geography: { geography_json }
}) => {
  const geographyFeature = geographyFeatureFromGeojson(geography_json)
  const [minLng, minLat, maxLng, maxLat] = bbox(geographyFeature)
  return [
    [minLng, maxLat], // SW
    [maxLng, minLat] // NE
  ]
}

export const getGeographiesBounds: (options: {
  geographies: Pick<Geography, 'geography_json'>[]
}) => LngLatBoundsLike = ({ geographies }) =>
  geographies.reduce<[[number, number], [number, number]]>(
    ([[globalMinLng, globalMaxLat], [globalMaxLng, globalMinLat]], { geography_json }) => {
      // FIXME: should this use getGeographyBounds?
      const geographyFeature = geographyFeatureFromGeojson(geography_json)
      const [localMinLng, localMinLat, localMaxLng, localMaxLat] = bbox(geographyFeature)
      return [
        [Math.min(localMinLng, globalMinLng), Math.max(localMaxLat, globalMaxLat)],
        [Math.max(localMaxLng, globalMaxLng), Math.min(localMinLat, globalMinLat)]
      ]
    },
    [
      [Number.MAX_SAFE_INTEGER, -Number.MAX_SAFE_INTEGER], // SW
      [-Number.MAX_SAFE_INTEGER, Number.MAX_SAFE_INTEGER] // NE
    ]
  )

export const getPointBounds: (options: { point: Point }) => LngLatBoundsLike = ({
  point: {
    coordinates: [lng, lat]
  }
}) => {
  return [
    [lng, lat], // SW
    [lng, lat] // NE
  ]
}

export type BoundsInputOptions<T> = {
  pois: T[]
  toLngLat: (poi: T) => { latitude: number; longitude: number } | undefined
}

export function getBounds<T>({ pois, toLngLat }: BoundsInputOptions<T>): LngLatBoundsLike {
  let minLng = Number.POSITIVE_INFINITY
  let minLat = Number.POSITIVE_INFINITY
  let maxLng = Number.NEGATIVE_INFINITY
  let maxLat = Number.NEGATIVE_INFINITY
  for (const poi of pois) {
    const lnglat = toLngLat(poi)
    if (lnglat) {
      const { latitude, longitude } = lnglat
      minLng = Math.min(minLng, longitude)
      minLat = Math.min(minLat, latitude)
      maxLng = Math.max(maxLng, longitude)
      maxLat = Math.max(maxLat, latitude)
    }
  }

  return [
    [minLng, maxLat], // SW
    [maxLng, minLat] // NE
  ]
}

export const makeViewState: (options: {
  dimensions: Dimensions
  bounds: LngLatBoundsLike
  padding?: number
  maxZoom?: number
}) => MapViewportState = ({ dimensions, bounds, padding, maxZoom }) => ({
  ...fitBounds({
    dimensions,
    bounds,
    padding: padding || 0,
    maxZoom: maxZoom || fitBoundsMaxZoom
  }),
  bearing: 0,
  pitch: 0
})

// This is a real gem of a function. Essentially you provide a
// grid of polygons that represent spatial bins (think hex, squares, triangles, whatever)
// and also provide a set of points that represent entities (think vehicles)
// and what you get in return is the original grid of polygons with data in the properties
// indicating the count of entities contained within the polygon.
// Adapted from https://gist.github.com/clhenrick/5787a12a8bf3b02821839e4f9556d997
export const spatialIntersect = ({
  grid,
  data
}: {
  grid: FeatureCollection<Polygon, GeoJsonProperties>
  data: FeatureCollection<Point, GeoJsonProperties>
}) => {
  if (grid.features.length === 0) {
    // eslint-disable-next-line no-console
    console.warn('Attempted to spatialIntersect on geojson without features.')
    return grid
  }

  // create the hexbin geometry for the given bbox and cell resolution
  // perform a "spatial join" on our grid geometry and our point data
  const collected = collect(grid, data, 'id', 'values')

  // get rid of polygons with no joined data, to reduce our final output file size
  collected.features = collected.features.filter(d => d.properties?.values.length)

  // count the number of instances per hexbin
  meta.propEach(collected, props => {
    if (props) {
      // eslint-disable-next-line no-param-reassign
      props.count = props.values.length
    }
  })

  // reduce our count values to a new array of numbers
  const reduced = meta.featureReduce(
    collected,
    (acc: number[], cur: Feature<Polygon, GeoJsonProperties>) => {
      return [...acc, cur.properties?.count]
    },
    []
  )

  // compute the ckMeans binning for data into 7 classes from reduced values
  const ck = simpleStats.ckmeans(reduced, Math.min(15, reduced.length))

  // tack on the bin number to our data, as well as its min and max values
  meta.propEach(collected, props => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    ck.forEach((bin: any, index: any) => {
      if (props && bin.indexOf(props.count) > -1) {
        // eslint-disable-next-line no-param-reassign
        props.bin = index
        // eslint-disable-next-line no-param-reassign
        props.binVal = d3Array.extent(bin)
      }
    })
  })

  // remove the "values" property from our hexBins as it's no longer needed
  meta.propEach(collected, props => {
    // eslint-disable-next-line no-param-reassign
    if (props) delete props.values
  })

  return collected
}

/* Builds a color ramp between 2 colors. This is very useful for mapbox color based step styles */
export const buildColorRamp = (
  startColor: string,
  endColor: string,
  startNumber: number,
  endNumber: number,
  totalCount: number
) => {
  const colorRamp = interpolateRgb(startColor, endColor)
  const stepSize = (endNumber - startNumber) / (totalCount - 1)
  return Array.from({ length: totalCount }, (v, k) => k / (totalCount - 1)).map((v, i) => [
    startNumber + stepSize * i,
    colorRamp(v)
  ])
}
