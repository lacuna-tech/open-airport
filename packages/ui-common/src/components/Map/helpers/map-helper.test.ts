import { Feature, MultiPolygon, Point, Polygon } from 'geojson'
import { geographyFeatureFromGeojson, getGeographiesBounds, getGeographyBounds, getPointBounds } from './map-helper'

const pointFeature: Feature<Point> = {
  type: 'Feature',
  properties: {},
  geometry: {
    type: 'Point',
    coordinates: [-122.3493, 47.6205]
  }
}

const pointBounds = [
  [-122.34970667259853, 47.62077411283809],
  [-122.34889332740148, 47.620225887161915]
]

const polygonFeature: Feature<Polygon> = {
  type: 'Feature',
  properties: {},
  geometry: {
    type: 'Polygon',
    coordinates: [
      [
        [-122.34934777021407, 47.62066174448532],
        [-122.3495677113533, 47.620473721969276],
        [-122.34936252236366, 47.62028569877702],
        [-122.34893471002579, 47.62035982338546],
        [-122.34900847077368, 47.620557789812416],
        [-122.34934777021407, 47.62066174448532]
      ]
    ]
  }
}

const polygonBounds = [
  [-122.3495677113533, 47.62066174448532],
  [-122.34893471002579, 47.62028569877702]
]

const multiPolygonFeature: Feature<MultiPolygon> = {
  type: 'Feature',
  properties: {},
  geometry: {
    type: 'MultiPolygon',
    coordinates: [
      [
        [
          [-122.34933704137802, 47.62080095302787],
          [-122.3493705689907, 47.620715981623896],
          [-122.34923109412193, 47.62066626424901],
          [-122.34921500086783, 47.62077021870622],
          [-122.34933704137802, 47.62080095302787]
        ]
      ],
      [
        [
          [-122.34921902418137, 47.62082897548188],
          [-122.34918147325516, 47.62070965396064],
          [-122.34907686710356, 47.62076750685342],
          [-122.34921902418137, 47.62082897548188]
        ]
      ]
    ]
  }
}

describe('geographyFeatureFromGeojson', () => {
  it('converts point geometries to circular polygon geometries', () => {
    const feature = geographyFeatureFromGeojson(
      {
        type: 'FeatureCollection',
        features: [pointFeature]
      },
      // Helps to insulate against default circular resolution changes.
      { pointGeometryCircleResolution: 64 }
    )

    // Circle is pretty data-heavy with 64 verticies, so we'll use a snapshot.
    expect(feature).toMatchSnapshot()
  })

  it('leaves Polygon & MultiPolygon features unchanged', () => {
    const feature = geographyFeatureFromGeojson({
      type: 'FeatureCollection',
      features: [polygonFeature, multiPolygonFeature]
    })

    expect(feature).toEqual({
      type: 'FeatureCollection',
      features: [polygonFeature, multiPolygonFeature]
    })
  })
})

describe('getGeographyBounds', () => {
  it('returns the correct bounds for a Polygon-based geography', () => {
    const bounds = getGeographyBounds({
      geography: {
        geography_json: {
          type: 'FeatureCollection',
          features: [polygonFeature]
        }
      }
    })

    expect(bounds).toEqual(polygonBounds)
  })

  it('returns the correct bounds for a Point-based geography (assume 100ft radius)', () => {
    const bounds = getGeographyBounds({
      geography: {
        geography_json: {
          type: 'FeatureCollection',
          features: [pointFeature]
        }
      }
    })

    expect(bounds).toEqual(pointBounds)
  })
})

describe('getGeographiesBounds', () => {
  it('returns the correct bounds for multiple Polygon-based geographies', () => {
    const bounds = getGeographiesBounds({
      geographies: [
        {
          geography_json: {
            type: 'FeatureCollection',
            features: [polygonFeature]
          }
        },
        {
          geography_json: {
            type: 'FeatureCollection',
            features: [multiPolygonFeature]
          }
        }
      ]
    })

    expect(bounds).toEqual([
      [-122.3495677113533, 47.62082897548188],
      [-122.34893471002579, 47.62028569877702]
    ])
  })

  it('returns the correct bounds for a mix of Polygon & Point-based geography (assume 100ft radius)', () => {
    const bounds = getGeographiesBounds({
      geographies: [
        {
          geography_json: {
            type: 'FeatureCollection',
            features: [pointFeature]
          }
        },
        {
          geography_json: {
            type: 'FeatureCollection',
            features: [multiPolygonFeature]
          }
        }
      ]
    })

    expect(bounds).toEqual([
      [-122.34970667259853, 47.62082897548188],
      [-122.34889332740148, 47.620225887161915]
    ])
  })
})

describe('getPointBounds', () => {
  it('returns zero-width bounds', () => {
    const bounds = getPointBounds({
      point: {
        type: 'Point',
        coordinates: [-122.3493, 47.6205]
      }
    })

    expect(bounds).toEqual([
      [-122.3493, 47.6205],
      [-122.3493, 47.6205]
    ])
  })
})
