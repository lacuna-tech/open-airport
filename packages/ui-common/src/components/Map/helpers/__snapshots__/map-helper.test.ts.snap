// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`geographyFeatureFromGeojson converts point geometries to circular polygon geometries 1`] = `
Object {
  "features": Array [
    Object {
      "geometry": Object {
        "coordinates": Array [
          Array [
            Array [
              -122.3493,
              47.62077411283809,
            ],
            Array [
              -122.34933986109316,
              47.62077279290294,
            ],
            Array [
              -122.3493793382962,
              47.62076884580941,
            ],
            Array [
              -122.34941805141634,
              47.62076230957076,
            ],
            Array [
              -122.34945562761972,
              47.62075324713547,
            ],
            Array [
              -122.34949170502236,
              47.62074174578105,
            ],
            Array [
              -122.34952593617528,
              47.62072791627341,
            ],
            Array [
              -122.34955799141099,
              47.62071189180004,
            ],
            Array [
              -122.34958756201826,
              47.62069382668732,
            ],
            Array [
              -122.34961436321538,
              47.62067389491419,
            ],
            Array [
              -122.34963813689279,
              47.62065228843653,
            ],
            Array [
              -122.34965865409869,
              47.62062921533848,
            ],
            Array [
              -122.34967571724408,
              47.620604897828365,
            ],
            Array [
              -122.34968916200525,
              47.62057957009869,
            ],
            Array [
              -122.34969885890655,
              47.620553476070604,
            ],
            Array [
              -122.3497047145668,
              47.620526867044816,
            ],
            Array [
              -122.34970667259853,
              47.6204999992814,
            ],
            Array [
              -122.3497047141508,
              47.62047313153178,
            ],
            Array [
              -122.34969885809058,
              47.62044652254689,
            ],
            Array [
              -122.34968916082067,
              47.62042042858521,
            ],
            Array [
              -122.34967571573635,
              47.620395100944904,
            ],
            Array [
              -122.34965865232583,
              47.620370783543684,
            ],
            Array [
              -122.34963813492286,
              47.62034771056987,
            ],
            Array [
              -122.34961436112413,
              47.62032610422702,
            ],
            Array [
              -122.34958755988605,
              47.62030617259408,
            ],
            Array [
              -122.34955798931972,
              47.62028810762155,
            ],
            Array [
              -122.34952593420537,
              47.620272083282984,
            ],
            Array [
              -122.34949170324946,
              47.62025825389958,
            ],
            Array [
              -122.349455626112,
              47.620246752654054,
            ],
            Array [
              -122.3494180502317,
              47.62023769030814,
            ],
            Array [
              -122.34937933748022,
              47.62023115413589,
            ],
            Array [
              -122.34933986067719,
              47.620227207083246,
            ],
            Array [
              -122.3493,
              47.620225887161915,
            ],
            Array [
              -122.34926013932282,
              47.620227207083246,
            ],
            Array [
              -122.34922066251978,
              47.62023115413589,
            ],
            Array [
              -122.3491819497683,
              47.62023769030814,
            ],
            Array [
              -122.34914437388801,
              47.620246752654054,
            ],
            Array [
              -122.34910829675054,
              47.62025825389958,
            ],
            Array [
              -122.34907406579464,
              47.620272083282984,
            ],
            Array [
              -122.34904201068026,
              47.62028810762155,
            ],
            Array [
              -122.34901244011397,
              47.62030617259408,
            ],
            Array [
              -122.34898563887589,
              47.62032610422702,
            ],
            Array [
              -122.34896186507714,
              47.62034771056987,
            ],
            Array [
              -122.34894134767418,
              47.620370783543684,
            ],
            Array [
              -122.34892428426365,
              47.620395100944904,
            ],
            Array [
              -122.34891083917934,
              47.62042042858521,
            ],
            Array [
              -122.34890114190944,
              47.62044652254689,
            ],
            Array [
              -122.34889528584921,
              47.62047313153178,
            ],
            Array [
              -122.34889332740148,
              47.6204999992814,
            ],
            Array [
              -122.34889528543322,
              47.620526867044816,
            ],
            Array [
              -122.34890114109346,
              47.620553476070604,
            ],
            Array [
              -122.34891083799475,
              47.62057957009869,
            ],
            Array [
              -122.34892428275593,
              47.620604897828365,
            ],
            Array [
              -122.34894134590131,
              47.62062921533848,
            ],
            Array [
              -122.34896186310722,
              47.62065228843653,
            ],
            Array [
              -122.34898563678463,
              47.62067389491419,
            ],
            Array [
              -122.34901243798176,
              47.62069382668732,
            ],
            Array [
              -122.34904200858902,
              47.62071189180004,
            ],
            Array [
              -122.34907406382472,
              47.62072791627341,
            ],
            Array [
              -122.34910829497764,
              47.62074174578105,
            ],
            Array [
              -122.3491443723803,
              47.62075324713547,
            ],
            Array [
              -122.34918194858368,
              47.62076230957076,
            ],
            Array [
              -122.34922066170381,
              47.62076884580941,
            ],
            Array [
              -122.34926013890686,
              47.62077279290294,
            ],
            Array [
              -122.3493,
              47.62077411283809,
            ],
          ],
        ],
        "type": "Polygon",
      },
      "properties": Object {},
      "type": "Feature",
    },
  ],
  "type": "FeatureCollection",
}
`;
